### PR TITLE
[WIP] cacher: unblock watchers stuck in draining after the client disconnected

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
it's a wip pr because i'd like to add more unit tests

xref: https://github.com/kubernetes/kubernetes/issues/121438#issuecomment-1776957860

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -1223,7 +1222,8 @@ func forgetWatcher(c *Cacher, w *cacheWatcher, index int, scope namespacedName,
 		// It's possible that the watcher is already not in the structure (e.g. in case of
 		// simultaneous Stop() and terminateAllWatchers(), but it is safe to call stopLocked()
 		// on a watcher multiple times.
-		c.watchers.deleteWatcher(index, scope, triggerValue, triggerSupported, c.stopWatcherLocked)
+		c.watchers.deleteWatcher(index, scope, triggerValue, triggerSupported)
+		c.stopWatcherLocked(w)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1226]: I'm assuming this is a cleanup change - if so, let's move it to a separate PR...

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1226]: opened https://github.com/kubernetes/kubernetes/pull/121615

**Comment by p0lyn0mial** [comment]
closing this PR based on https://github.com/kubernetes/kubernetes/pull/121568#discussion_r1377339162

/close
